src/nmod_poly/divrem_basecase.c

```diff
@@ -101,10 +101,15 @@ void _nmod_poly_divrem_q1_preinv1(mp_ptr Q, mp_ptr R,
         {
             for (i = 1; i < lenB - 1; i++)
             {
-                t = A[i];
-                NMOD_ADDMUL(t, q1, B[i - 1], mod);
-                NMOD_ADDMUL(t, q0, B[i], mod);
-                R[i] = t;
+                umul_ppmm(t1, t0, q1, B[i - 1]);
+                umul_ppmm(s1, s0, q0, B[i]);
+                add_sssaaaaaa(t2, t1, t0, 0,  t1, t0, 0, 0, A[i]);
```

Comment:
If you have full multiplication, i.e. $p_1 B^1 + p_0 = a_0 \cdot b_0$, then $p_1 \le B - 2$, so this can always allow an extra carry without overflowing.

Reply:
You can check this with $a_0 = b_0 = B - 1$.

Reply:
Yes!